src/util/codec/bytes.rs

```diff
@@ -199,11 +214,75 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
                 *k = !*k;
             }
         }
+        // data will point to following unencoded bytes, maybe timestamp
         *data = &data[offset..];
         return Ok(key);
     }
 }
 
+/// `decode_bytes_in_place` decodes bytes which is encoded by `encode_bytes` before just in place without malloc.
+/// Please use this instead of `decode_bytes` if possible.
+pub fn decode_bytes_in_place(data: &mut Vec<u8>, desc: bool) -> Result<()> {
+    let mut write_offset = 0;
+    let mut read_offset = 0;
+    loop {
+        let marker_offset = read_offset + ENC_GROUP_SIZE;
+        // everytime make ENC_GROUP_SIZE + 1 elements as a decode unit
+        let next_offset = marker_offset + 1;
+        if next_offset > data.len() {
+            return Err(Error::unexpected_eof());
+        };
+
+        unsafe {
+            // it is semantically equivalent to C's memmove()
+            // and the src and dest may overlap
+            // if src == dest do nothing
+            ptr::copy(
+                data.as_ptr().offset(read_offset as isize),
+                data.as_mut_ptr().offset(write_offset as isize),
+                ENC_GROUP_SIZE,
+            );
+        }
+        write_offset += ENC_GROUP_SIZE;
+        read_offset = next_offset;
+
+        unsafe {
+            // the last byte in decode unit is for marker which indicates pad size
+            // for the sake of performancce, use get_unchecked() to save bound check
```

Comment:
I don't think bound check exists anymore as L232 has proven it to the compiler.

Reply:
maybe it can not make sure `data.len()` whether change or not, and from bench result showed, uncheck does make sense.

Reply:
Seems `+ 1` confuse compiler. Using `marker_offset >= data.len()` can omit the bound check.

Reply:
you are right, `marker_offset >= data.len()` makes sense.